src/Tools/dotnet-counters/CounterMonitor.cs

```diff
@@ -834,85 +540,20 @@ private static void ParseCounterProvider(string providerText, CounterSet counter
-                    metrics.Append(string.Join(',', providerCounters));
-                }
-            }
-
-            // Shared Session Id was added in 8.0 - older runtimes will not properly support it.
-            _sessionId = Guid.NewGuid().ToString();
-            if (_diagnosticsClient.GetProcessInfo().TryGetProcessClrVersion(out Version version))
-            {
-                _sessionId = version.Major >= 8 ? SharedSessionId : _sessionId;
-            }
-
-            EventPipeProvider metricsEventSourceProvider =
-                new("System.Diagnostics.Metrics", EventLevel.Informational, TimeSeriesValues,
-                    new Dictionary<string, string>()
-                    {
-                        { "SessionId", _sessionId },
-                        { "Metrics", metrics.ToString() },
-                        { "RefreshInterval", _interval.ToString() },
-                        { "MaxTimeSeries", _maxTimeSeries.ToString() },
-                        { "MaxHistograms", _maxHistograms.ToString() },
-                        { "ClientId", _clientId  }
-                    }
-                );
-
-            return eventCounterProviders.Append(metricsEventSourceProvider).ToArray();
-        }
-
-        private bool Filter(string meterName, string instrumentName)
-        {
-            return _counterList.GetCounters(meterName).Contains(instrumentName) || _counterList.IncludesAllCounters(meterName);
-        }
+                ProviderName = provider,
+                CounterNames = _counterList.GetCounters(provider).ToArray()
+            }).ToArray();
 
-        private Task<int> Start()
+        private async Task<ReturnCode> Start(MetricsPipeline pipeline, CancellationToken token)
         {
-            EventPipeProvider[] providers = GetEventPipeProviders();
             _renderer.Initialize();
-
-            Task monitorTask = new(() => {
+            Task monitorTask = new(async () => {
                 try
                 {
-                    _session = _diagnosticsClient.StartEventPipeSession(providers, false, 10);
-                    if (_resumeRuntime)
-                    {
-                        try
-                        {
-                            _diagnosticsClient.ResumeRuntime();
-                        }
-                        catch (UnsupportedCommandException)
-                        {
-                            // Noop if the command is unknown since the target process is most likely a 3.1 app.
-                        }
-                    }
-                    EventPipeEventSource source = new(_session.EventStream);
-                    source.Dynamic.All += DynamicAllMonitor;
-                    _renderer.EventPipeSourceConnected();
-                    source.Process();
+                    await (await pipeline.StartAsync(token).ConfigureAwait(false)).ConfigureAwait(false);
```

Comment:
Nit: Is this Task\<Task\>? Assuming it is or something similar how about split with a local variable so we limit to one await per line.